target/sh.c

```diff
@@ -58,32 +68,59 @@ static void sh_emit_inst(Inst* inst) {
     break;
 
   case ADD:
-    emit_line("%s=$(( ($%s + %s) & " UINT_MAX_STR " ))",
+    emit_line(": $(( %s = (($%s + %s) & " UINT_MAX_STR ") ))",
               reg_names[inst->dst.reg],
               reg_names[inst->dst.reg], sh_src_str(inst));
     break;
 
   case SUB:
-    emit_line("%s=$(( ($%s - %s) & " UINT_MAX_STR "))",
+    emit_line(": $(( %s = (($%s - %s) & " UINT_MAX_STR ") ))",
               reg_names[inst->dst.reg],
               reg_names[inst->dst.reg], sh_src_str(inst));
     break;
 
   case LOAD:
-    emit_line("eval z=\\$m%s", sh_src_str(inst));
-    emit_line("%s=$(( $z + 0 ))", reg_names[inst->dst.reg]);
+    emit_line(": $(( %s = m%s + 0 ))", reg_names[inst->dst.reg], sh_src_str(inst));
     break;
 
   case STORE:
-    emit_line("eval m%s=$%s", sh_src_str(inst), reg_names[inst->dst.reg]);
+    emit_line(": $(( m%s = $%s ))", sh_src_str(inst), reg_names[inst->dst.reg]);
     break;
 
   case PUTC:
-    emit_line("t=$((%s&255))", sh_src_str(inst));
-    emit_line("printf \"\\\\$(printf '%%03o' $t)\"");
+    emit_line(": $(( t = (%s & 255) ))", sh_src_str(inst));
+    emit_line("printf \\\\$((t/64))$((t/8%%8))$((t%%8))");
     break;
 
   case GETC:
+#ifdef POSIX_SHELL
+    // The POSIX standard doesn't support read -n1, so we read a full line and buffer it.
+    // The GETC state machines can be in 3 states:
+    //   Non-empty GETC_BUF buffer: take first char from buffer
+    //   Empty GETC_BUF buffer AND GETC_BUF_ENDING = -1: need to read a new line
+    //   Empty GETC_BUF buffer AND GETC_BUF_ENDING != 0: Need to output the end of line char
+    emit_line("while :; do"); // while to allow breaking out early
+    emit_line(" if [ -z \"$GETC_BUF\" ]; then");
+    emit_line("  if [ $GETC_BUF_ENDING != -1 ]; then");
+    emit_line("   %s=$GETC_BUF_ENDING", reg_names[inst->dst.reg]);
+    emit_line("   : $(( GETC_BUF_ENDING = -1 ))");
+    emit_line("   break");
+    emit_line("  fi");
+    emit_line("  GETC_BUF_ENDING=10");
+    emit_line("  IFS= read -r GETC_BUF || GETC_BUF_ENDING=0");
+    emit_line("  if [ -z \"$GETC_BUF\" ]; then");
+    emit_line("   : $(( %s = GETC_BUF_ENDING ))", reg_names[inst->dst.reg]);
+    emit_line("   : $(( GETC_BUF_ENDING = -1 ))");
+    emit_line("   break");
+    emit_line("  fi");
+    emit_line(" fi");
+    emit_line(" GETC_BUF2=\"${GETC_BUF#?}\"");
+    emit_line(" t=\"${GETC_BUF%%\"$GETC_BUF2\"}\"");
+    emit_line(" GETC_BUF=\"$GETC_BUF2\"");
+    emit_line(" %s=$(printf '%%d' \"'$t'\")", reg_names[inst->dst.reg]);
+    emit_line(" break");
+    emit_line("done");
```

Comment:
This opcode implementation is a bit long and could probably be factored out as a separate function. Let me know if you want me to do that.